src/devices/cellular.c

```diff
@@ -0,0 +1,534 @@
+        struct serial_buffer *sb = (struct serial_buffer*) config;
+
+        telemetry_info.status = TELEMETRY_STATUS_IDLE;
+        telemetry_info.active_since = 0;
+
+        /* Only disconnect if the socket is >= 0 */
+        if (telemetry_info.socket < 0) {
+                pr_info("[cell] Already disconnected\r\n");
+                return 0;
+        }
+
+        return methods ? methods->close_telem_connection(
+                sb, &cell_info, &telemetry_info) : 0;
+}
+
+static void reset_modem()
+{
+        pr_info("[cell] Resetting modem...\r\n");
+
+        /*
+         * SIM900 requres at least 1 second of pull for powerdown
+         * and then 1 second of non-pull for powerup.  This however
+         * can screw with Ublox-Sara, which needs to autobaud very
+         * shortly (< 3sec) after powerup.  This balance seems to work
+         * "Well enough"(tm).
+         */
+        cell_pwr_btn(true);
+        delayMs(1100);
+        cell_pwr_btn(false);
+        delayMs(1100);
+
+        at_cfg = get_safe_at_config();
+}
+
+/**
+ * Puts the device in a state that is ready to use.
+ */
+static int cellular_init_modem(struct serial_buffer *sb)
+{
+        reset_modem();
+
+        if (!autobaud_modem(sb, 5, 200))
+                return DEVICE_INIT_FAIL;
+
+        if(!gsm_set_echo(sb, false))
+                return DEVICE_INIT_FAIL;
+
+        return DEVICE_INIT_SUCCESS;
+}
+
+static void print_registration_failure()
+{
+        pr_info("[cell] Network registration failed: ");
+        switch(cell_info.net_status) {
+        case CELLULAR_NETWORK_SEARCHING:
+                pr_info("Network not found");
+                break;
+        case CELLULAR_NETWORK_DENIED:
+                pr_info("Access denied.  Call provider.");
+                break;
+        default:
+                pr_info("Unknown reason. ");
+                pr_info_int_msg("Message code ", cell_info.net_status);
+                return;
+        }
+        pr_info("\r\n");
+}
+
+static bool auth_telem_stream(struct serial_buffer *sb,
+                              const TelemetryConfig *tc)
+{
+        serial_buffer_reset(sb);
+
+        const char *deviceId = tc->telemetryDeviceId;
+
+        Serial *serial = sb->serial;
+        json_objStart(serial);
+        json_objStartString(serial, "auth");
+        json_string(serial, "deviceId", deviceId, 1);
+        json_int(serial, "apiVer", API_REV, 1);
+        json_string(serial, "device", DEVICE_NAME, 1);
+        json_string(serial, "ver", VERSION_STR, 1);
+        json_string(serial, "sn", cpu_get_serialnumber(), 0);
+        json_objEnd(serial, 0);
+        json_objEnd(serial, 0);
+        serial->put_c('\n');
+
+        pr_debug_str_msg("sending auth- deviceId: ", deviceId);
+
+        /* Parse the incoming JSON */
+        jsmn_parser parser;
+        jsmntok_t toks[AUTH_MAX_JSMN_TOKENS];
+
+        for (size_t tries = 20; tries; --tries) {
```

Comment:
I don't recall why 20 off-hand.  Probably a left-over that should get updated.  If the connection drops at this point we will never auth and the modem will reset and try again.  I can try to optimize a bit here but we need to be careful otherwise we can find ourselves in a loop where we never auth.